strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,353 @@
+    number of modes due to the mode-shifting occuring in TDMProgram.
+
+    Args:
+        all_samples (dict[int, list]): the raw measured samples
+        modes (Sequence[int]): the modes that are measured in the circuit
+        N (Sequence[int]): the number of used modes per belt/spatial modes
+
+    Returns:
+        new_samples (dict[int, list]): the re-shaped samples
+    """
+    num_of_values = len([i for j in all_samples.values() for i in j])
+    mode_order = _get_mode_order(num_of_values, N)
+
+    # go backwards through all_samples and add them into the correct mode
+    new_samples = dict()
+    for i, m in enumerate(mode_order):
+        idx = modes[i % len(N)]
+        if idx not in new_samples:
+            new_samples[idx] = []
+        new_samples[idx].append(all_samples[m].pop(0))
+    return new_samples
+
+
+class TDMProgram(sf.Program):
+    r"""Represents a photonic quantum circuit in the time domain encoding.
+
+    The ``TDMProgam`` class provides a context manager for easily defining
+    a single time-bin of the time domain algorithm. As with the standard
+    :class:`~.Program`, Strawberry Fields operations are appended to the
+    time domain program using the Python-embedded Blackbird syntax.
+
+    Once created, time domain programs can be executed on Strawberry Fields'
+    suite of built-in simulators in an efficient manner, or submitted
+    to be executed on compatible hardware.
+
+    Args:
+        N (int or Sequence[int]): If an integer, the number of concurrent (or 'alive')
+            modes in each time bin. Alternatively, a sequence of integers
+            may be provided, corresponding to
+        name (str): the program name (optional)
+
+    **Example**
+
+    Below, we create a time domain program with 2 concurrent modes:
+
+    >>> import strawberryfields as sf
+    >>> from strawberryfields import ops
+    >>> prog = sf.tdm.TDMProgram(N=2)
+
+    Once created, we can construct the program using the ``prog.context()``
+    context manager.
+
+    >>> with prog.context([1, 2], [3, 4], copies=3) as (p, q):
+    ...     ops.Sgate(0.7, 0) | q[1]
+    ...     ops.BSgate(p[0]) | (q[0], q[1])
+    ...     ops.MeasureHomodyne(p[1]) | q[0]
+
+    If we print out this program, we see that the time domain program
+    has automated the process of repeating the single time-bin sequence
+    constructed above:
+
+    >>> prog.print()
+    Sgate(0.32, 0) | (q[1])
+    BSgate(1, 0) | (q[0], q[1])
+    MeasureHomodyne(3) | (q[0])
+    Sgate(0.32, 0) | (q[0])
+    BSgate(2, 0) | (q[1], q[0])
+    MeasureHomodyne(4) | (q[1])
+    Sgate(0.32, 0) | (q[1])
+    BSgate(1, 0) | (q[0], q[1])
+    MeasureHomodyne(3) | (q[0])
+    Sgate(0.32, 0) | (q[0])
+    BSgate(2, 0) | (q[1], q[0])
+    MeasureHomodyne(4) | (q[1])
+    Sgate(0.32, 0) | (q[1])
+    BSgate(1, 0) | (q[0], q[1])
+    MeasureHomodyne(3) | (q[0])
+    Sgate(0.32, 0) | (q[0])
+    BSgate(2, 0) | (q[1], q[0])
+    MeasureHomodyne(4) | (q[1])
```

Comment:
Do we want to update the print as well to print something that makes more sense to the user? It feels like the `TDMProgram` is trying to copy the hardware, and thus "hide" the fact that it actually shifts the modes around. This could make users a bit confused if expecting the circuit to look like the one they've defined, as well as having `all_samples` being measured on fewer/different modes than those that are printed using `prog.print()`. :thinking:

Reply:
I kind of like this print...

Reply:
I can see how it might be confusing though! Perhaps `prog.print()` could receive a keyword argument switching the display mode. E.g., `prog.print()` displays it without the shift, `prog.print(simulation_mode=True)` displays it with the shift?

Reply:
No objections against additional print options but, personally, I like the way the circuit is printed in the above example. In my opinion, it does not hide the fact that the modes are shifted but, rather, just reflects that qumodes are being indexed completely differently in a TDMProgram (indices corresponding to the "location" in the TDM setup instead of sticking to their qumode throughout the interferometer). Once the users have accepted this paradigm, then I think the above representation will seem intuitive to them. @thisac, do you think an explicit mention of the shift in the print would help the user to see what's going on? Like for instance

```
>>> prog.print()
Sgate(0.32, 0) | (q[1])
BSgate(1, 0) | (q[0], q[1])
MeasureHomodyne(3) | (q[0])
Shift | (q[0],q[1])
Sgate(0.32, 0) | (q[0])
BSgate(2, 0) | (q[1], q[0])
MeasureHomodyne(4) | (q[1])
Shift | (q[0],q[1])
Sgate(0.32, 0) | (q[1])
BSgate(1, 0) | (q[0], q[1])
MeasureHomodyne(3) | (q[0])
Shift | (q[0],q[1])
Sgate(0.32, 0) | (q[0])
...
...
```

Reply:
![qumode conveyor belt (1)](https://user-images.githubusercontent.com/62264296/92598032-38ba0800-f276-11ea-9ac4-60a2626ba9b9.png)

I know it's a bit stupid, but this is how I picture a TDMProgram. The gates and the qumode indices are defined by the static upper disc. The actual light modes are written on the lower disc, rotating from one window to another. Not too relevant for this discussion, but found it cute and wanted to share. :)

Reply:
> Once the users have accepted this paradigm, then I think the above representation will seem intuitive to them.

I might be missing part of the idea here, but I sort of thought that part of the reason for having a `TDMProgram` was to not make the user having to know about the inner workings of it at all. :thinking: 

> @thisac, do you think an explicit mention of the shift in the print would help the user to see what's going on? Like for instance

That could help a bit perhaps, but I don't think it's necessary. I kind of like the suggestion from @josh146; having both prints available. But I don't really consider this an important issue, so feel free leave it as it is. :slightly_smiling_face: